scylla.go

```diff
@@ -331,3 +394,259 @@ func (p *scyllaConnPicker) randomConn() *Conn {
 	}
 	return nil
 }
+
+type scyllaDialerExt struct {
+	net.Dialer
+}
+
+var _ DialerExt = (*scyllaDialerExt)(nil)
+
+func wrapScyllaDialerExt(d *net.Dialer) *scyllaDialerExt {
+	return &scyllaDialerExt{*d}
+}
+
+func (spd *scyllaDialerExt) DialContextWithSourcePort(ctx context.Context, sourcePort uint16, network, addr string) (net.Conn, error) {
+	localAddr, err := net.ResolveTCPAddr(network, fmt.Sprintf(":%d", sourcePort))
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that we are copying the dialer struct, not just a pointer
+	var portDialer net.Dialer = spd.Dialer
+	portDialer.LocalAddr = localAddr
+
+	return portDialer.DialContext(ctx, network, addr)
+}
+
+type scyllaShardAwarePortDialer struct {
+	innerDialer DialerExt
+
+	scConnParams *scyllaConnParams
+}
+
+func wrapScyllaShardAwarePortDialer(innerDialer Dialer, scConnParams *scyllaConnParams) Dialer {
+	innerDialerExt, isExt := innerDialer.(DialerExt)
+	if !isExt {
+		return innerDialer
+	}
+
+	if !scConnParams.usesShardAwarePort() {
+		return innerDialer
+	}
+
+	return &scyllaShardAwarePortDialer{
+		innerDialer:  innerDialerExt,
+		scConnParams: scConnParams,
+	}
+}
+
+func (ssapd *scyllaShardAwarePortDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
```

Comment:
My take it that the whole DialerExt and scyllaDialerExt functionality could be squashed to this type making it more concise and easier to read in the end.